src/trace/trace.cpp

```diff
@@ -328,46 +343,56 @@ void Trace::add_lo2s_property(const std::string& name, const std::string& value)
     system_tree_node_properties_.emplace(system_tree_root_node_, intern(property_name),
                                          otf2::attribute_value{ intern(value) });
 }
-
+// As pid/tid pairs are unique in this context, create only one writer/location per pair
 otf2::writer::local& Trace::sample_writer(pid_t pid, pid_t tid)
 {
     auto name = (boost::format("thread %d") % tid).str();
-    auto location =
-        locations_.emplace(location_ref(), intern(name), location_groups_process_.at(pid),
-                           otf2::definition::location::location_type::cpu_thread);
-    process_comm_groups_.at(pid).add_member(location);
-    return archive()(location);
-}
+    auto location = thread_locations_.emplace(
+        std::piecewise_construct, std::forward_as_tuple(tid),
+        std::forward_as_tuple(location_ref(), intern(name), location_groups_process_.at(pid),
+                              otf2::definition::location::location_type::cpu_thread));
+    process_comm_groups_.at(pid).add_member(location.first->second);
 
+    return archive()(location.first->second);
+}
+// As CPU IDs are unique in this context, create only one writer/location per
+// CPU IDs
 otf2::writer::local& Trace::cpu_writer(int cpuid)
 {
-    auto name = intern((boost::format("cpu %d") % cpuid).str());
+    auto name = (boost::format("cpu %d") % cpuid).str();
+    auto intern_name = intern(name);
+
     auto r_group = location_groups_cpu_.emplace(
         std::piecewise_construct, std::forward_as_tuple(cpuid),
-        std::forward_as_tuple(location_group_ref(), name,
+        std::forward_as_tuple(location_group_ref(), intern_name,
                               otf2::definition::location_group::location_group_type::unknown,
                               system_tree_cpu_nodes_.at(cpuid)));
     const auto& group = r_group.first->second;
 
-    auto location = locations_.emplace(location_ref(), name, group,
-                                       otf2::definition::location::location_type::cpu_thread);
-    return archive()(location);
+    auto location = cpu_locations_.emplace(
+        std::piecewise_construct, std::forward_as_tuple(cpuid),
+        std::forward_as_tuple(location_ref(), intern_name, group,
+                              otf2::definition::location::location_type::cpu_thread));
+    return archive()(location.first->second);
 }
 
+// As pid/tid pairs are unique in this context, create only one
```

Comment:
This comment should be inside the function

Also, `tid`alone is already unique.